sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,1013 @@
+        this parameter, using brute force.
+
+    targets : array, shape (n_samples, n_neighbors), optional (default=None)
+        The target neighbors of each sample, namely the `n_neighbors` nearest
+        neighbors that belong to the same class. If None (default), the target
+        neighbors will be computed on the fly using the euclidean metric.
+
+    max_constraints : int, optional (default=500000)
+        Maximum number of constraints to enforce per iteration.
+
+    use_sparse : bool, optional (default=True)
+        Whether to use a sparse matrix (default) or a dense matrix for the
+        impostor-pairs storage. Using a sparse matrix, the (squared)
+        distance to impostors is computed twice, but it is faster for larger
+        data sets than using a dense matrix. With a dense matrix, the unique
+        impostor pairs have to be identified explicitly.
+
+    max_iter : int, optional (default=50)
+        Maximum number of iterations in the optimization.
+
+    tol : float, optional (default=1e-5)
+        Convergence tolerance for the optimization.
+
+    max_corrections : int, optional (default=100)
+        The maximum number of variable metric corrections
+        used to define the limited memory matrix. (The limited memory BFGS
+        method does not store the full hessian but uses this many terms in an
+        approximation to it.)
+
+    callback : callable, optional (default=None)
+        If not None, this function is called after every iteration of the
+        optimizer taking as arguments the current solution and the number of
+        iterations.
+
+    verbose : int, optional (default=0)
+        If 0, no progress messages will be printed.
+        If 1, progress messages will be printed to stdout.
+        If >1, progress messages will be printed and the ``iprint``
+        parameter of :meth:`fmin_l_bfgs_b` of `scipy.optimize` will be set to
+        verbose - 2.
+
+    random_state : int or numpy.RandomState or None, optional (default=None)
+        A pseudo random number generator object or a seed for it if int.
+
+    n_jobs : int, optional (default=1)
+        The number of parallel jobs to run for neighbors search.
+        If ``-1``, then the number of jobs is set to the number of CPU cores.
+        Doesn't affect :meth:`fit` method.
+
+    Attributes
+    ----------
+    transformation_ : array, shape (n_features_out, n_features).
+        The linear transformation learned during fitting.
+
+    n_neighbors_ : int
+        The provided n_neighbors is decreased if it is greater than or equal
+        to  min(number of elements in each class).
+
+    classes_inverse_non_singleton_ : array, shape (n_classes_non_singleton,)
+        The appearing classes that have more than one sample, transformed to
+        be integers within the range(0, n_classes).
+
+    n_funcalls_ : int
+        Counts the number of times the loss and the gradient are computed.
```

Comment:
is it really needed to store this?

Reply:
well think in the long run. We'll have to modify this code, for example propose a more efficient stochastic solvers so by exposing all this you'll create more work for refactoring and eventual deprecations

Reply:
I think in the case of LMNN, L-BFGS has been shown to work best after other optimizers were tried first. But if you insist, I can take it out.

Reply:
if you have really large n_samples stochastic approach will be necessary. Yes please remove it.